firebase-perf/src/main/java/com/google/firebase/perf/session/SessionManager.java

```diff
@@ -125,6 +125,17 @@ public void updatePerfSession(ApplicationProcessState currentAppState) {
     startOrStopCollectingGauges(currentAppState);
   }
 
+  /**
+   * Initial start of gauge collection. This should be called in ContentProvider.attachInfo during
+   * cold-start, because we want to start gauge collection as early as possible. This assumes {@link
+   * PerfSession} was already initialized a moment ago by getInstance(). Unlike updatePerfSession,
+   * this does not reset the perfSession.
+   */
+  public void initializeGaugeCollection() {
+    logGaugeMetadataIfCollectionEnabled(ApplicationProcessState.FOREGROUND);
+    startOrStopCollectingGauges(ApplicationProcessState.FOREGROUND);
```

Comment:
This API `startOrStopCollectingGauges` does not check if we are already collecting gauges or not. So, would there be a possibility where we might be triggering collection from two separate timers? For eg:

calling 
```
startOrStopCollectingGauges(ApplicationProcessState.FOREGROUND)
startOrStopCollectingGauges(ApplicationProcessState.FOREGROUND)
```

this could trigger two collections initiated at the same time. Can we make this atomic to ensure that we do not collect gauges twice at the same time?

Reference: 
https://github.com/firebase/firebase-android-sdk/blob/master/firebase-perf/src/main/java/com/google/firebase/perf/session/SessionManager.java#L160

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/CpuGaugeCollector.java#L116

Reply:
I'm guessing that the following blocks should take care of that:

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/CpuGaugeCollector.java#L126-L130

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/MemoryGaugeCollector.java#L83-L89

But, still can you confirm if we are not opening up two separate schedulers? Once confirmed, we are good to go!

Reply:
Yes, in fact, it will stop earlier than those 2 block of code. In `GaugeManager.startCollectingGauges(...)`, it checks `this.sessionId`, if it is not null, then call `stopCollectingGauges()`. It isn't obvious when you look at the variable name and type but `GaugeManager.sessionId` acts as a flag for if we are already collecting gauges.

[This line](https://github.com/firebase/firebase-android-sdk/blob/c91adbebfbc8c5366edef712ba66293eebdb5338/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/GaugeManager.java#L129) sets `GaugeManager.sessionId` when gauge collection starts.

[This line](https://github.com/firebase/firebase-android-sdk/blob/c91adbebfbc8c5366edef712ba66293eebdb5338/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/GaugeManager.java#L120) stops gauge collection because `sessionId` was already set. `stopCollectingGauges()` also [resets `sessionId` back to null](https://github.com/firebase/firebase-android-sdk/blob/c91adbebfbc8c5366edef712ba66293eebdb5338/firebase-perf/src/main/java/com/google/firebase/perf/session/gauges/GaugeManager.java#L216).

Reply:
Thanks for confirming @leotianlizhan. Interesting part is the GaugeManager reseting sessionId to null when stopCollectingGauges(). Ideally gauge manager should not worry about reseting the sessionID to null. That is a refactor/cleanup for later.

Reply:
The concern is the source of truth. GaugeManager would have an instance of sessionID which could be null, but the sessionManager would have an instance of sessionID which is not null. This is a recipe of confusion and possibly inconsistencies. As much as possible, we should avoid multiple source of truths. So the callout for the cleanup.